matrix_sdk/src/client.rs

```diff
@@ -1817,14 +1894,13 @@ impl Client {
 
             last_sync_time = Some(now);
 
-            sync_settings = SyncSettings::new().timeout(DEFAULT_SYNC_TIMEOUT).token(
+            sync_settings.token = Some(
                 self.sync_token()
                     .await
                     .expect("No sync token found after initial sync"),
             );
-            if let Some(f) = filter {
-                sync_settings = sync_settings.filter(f);
-            }
+            // FIXME: this should be direclty the default value for the timeout
+            sync_settings.timeout = sync_settings.timeout.or(Some(DEFAULT_SYNC_TIMEOUT));
```

Comment:
I think the long timeout should be user settable, if any :). For the first sync IMO you want pretty much always to receive the response as quickly as possible. I assumed that `sync` returns immediately when there isn't a `since` token set, but the docs say only that a sync request returns immediately (ignoring the timeout) when `full_state` is `true`.

Reply:
I had some discussion in #matrix-client-developers and from that i took this conclusions.

When no `since` or `full_state=true` the timeout is ignored and the homeserver will send the response as soon as possible, but an initial sync (probably also other syncs) can take a really long time to be prepend and actually be send since they may include a lot of data. Hence we can expose one single `timeout` for long-polling via `SyncSettings`.

The question is what request timeout can we use? This is a little complicated since we don't know how long it takes the server to prepare and send the response. I think the best option here are to only use a `connect_timeout` (the duration a request can have for establishing a server connection) or/and set a really long-timeout (>10 Min) request timeout.

Reply:
> The question is what request timeout can we use? This is a little complicated since we don't know how long it takes the server to prepare and send the response. I think the best option here are to only use a `connect_timeout` (the duration a request can have for establishing a server connection) or/and set a really long-timeout (>10 Min) request timeout.

I'm not sure I agree here. Servers handle cancellations on the `/sync` endpoint just fine, other SDKs cancel quite aggressively and retry the sync. If we're waiting 10 minutes for a sync the client might send messages just fine but appear to be stuck when receiving them.

What's wrong with the current approach of `sync_timeout` + N (N was 15 seconds?).

Reply:
> I'm not sure I agree here. Servers handle cancellations on the /sync endpoint just fine, other SDKs cancel quite aggressively and retry the sync. If we're waiting 10 minutes for a sync the client might send messages just fine but appear to be stuck when receiving them.

If this is true, than we can have a short timeout like for others. Which is good :)


> What's wrong with the current approach of sync_timeout + N (N was 15 seconds?).

Works for me. I would do `N = sync_timeout + request_config.timeout`  (request_config.timeout defaults to 10s). So that we don't have multiple default timeouts that are essentially the same and the user can set the timeout. Btw the code already has this behavior.

Reply:
Let's stick with it in that case.

Reply:
Wait, is this is going to misbehave if someone sets the timeout to 0?

Reply:
which timeout?

Reply:
If the request timeout is 0 then all requests will fail since they didn't conclude within 0s. we could eventually make it disable the timeout, but it wouldn't be really correct. 
If the sync timeout is set to 0, then long polling is disabled.

Reply:
If our sync timeout is 30 and the request timeout is 0 the request will not fail immediately but will fail after `sync_timeout` + 0 seconds. Perhaps we should clamp the request timeout to 5 seconds minimum so people can't missconfigure this?

Reply:
> If our sync timeout is 30 and the request timeout is 0 the request will not fail immediately but will fail after sync_timeout + 0 seconds. 
Right, i missed that case.

>Perhaps we should clamp the request timeout to 5 seconds minimum so people can't missconfigure this?

I think that's the bad kind of baby sitting and may create more confusion. If we want to prevent people from missconfiguring the sdk we should remove the configuration option.

Reply:
> I think that's the bad kind of baby sitting and may create more confusion. If we want to prevent people from missconfiguring the sdk we should remove the configuration option.

Fine, let's leave it as is.